### PR TITLE
[Perf] Support TeaCache for MammothModa2 

### PR DESCRIPTION
**Pull request by nagisa-kunhah**
## Purpose
part of #1217 
Related: #7106 

## Implementation Approach

The implementation follows the existing diffusion cache backend pattern and keeps model-specific behavior inside the MammothModa2 stage path.

- Add `cache_backend` and `cache_config` to the per-stage model/engine config so cache settings can reach non-standard generation stages that still execute diffusion-style DiT workloads.
- Preserve top-level cache flags for multi-stage deployments by forwarding them into cache-capable stage `engine_args` when the stage does not already define an explicit cache setting. Stage-local config remains higher priority and is not overwritten.
- Keep the TeaCache hook implementation centralized in the existing TeaCache backend. MammothModa2 continues to use the existing custom enabler:

```text
TeaCacheBackend -> CUSTOM_TEACACHE_ENABLERS["MammothModa2DiTPipeline"]
```

- Enable the cache from `MammothModa2ForConditionalGeneration` only when the active stage is `model_stage="dit"` and `cache_backend` is explicitly set. This avoids adding MammothModa2-specific cache logic to the shared `GPUGenerationModelRunner`.
- Store the enabled cache backend on `MammothModa2DiTPipeline`, then refresh it inside the DiT forward path using the request's `num_inference_steps`. This mirrors the standard diffusion worker lifecycle: enable once during model setup, refresh once before each generation.
- Pass `teacache_branch="positive"` and `teacache_branch="negative"` into the MammothModa2 transformer calls so TeaCache can maintain separate cache state for CFG positive and negative branches.

## Test Plan

Tested with the local MammothModa2 checkpoint:

```text
/root/autodl-tmp/models/MammothModa2-Preview
```

GPU:

```text
NVIDIA H20, 97871 MiB
```

All image e2e runs used:

```text
height=1024
width=1024
seed=142
text_guidance_scale=4.0
cfg_range=[0.0, 1.0]
num_inference_steps=50
```

The first 4-prompt acceleration run used the old Lumina-bootstrap MammothModa2 coefficients with `TeaCache rel_l1_thresh=0.1`. Visual inspection showed visible drift on `nyc_motorcycle` and `hf_city_zh`, so I collected MammothModa2 full-compute traces, fitted new polynomial coefficients on a larger 12-prompt calibration set, validated the candidates on 4 holdout prompts, and reran the e2e suite with the new default coefficients.

The expanded e2e suite used 4 prompts:

| Slug | Prompt |
|---|---|
| `nyc_motorcycle` | `A stylish woman riding a motorcycle through rainy New York City at night, cinematic movie poster style, neon reflections, sharp focus.` |
| `cat_laptop` | `A fluffy orange cat sitting on a laptop keyboard in a cozy home office, warm desk lamp, shallow depth of field, detailed fur.` |
| `futuristic_city` | `A futuristic coastal city at sunset with glass skyscrapers, flying transit vehicles, sailboats on calm water, detailed architecture, soft pink sky.` |
| `hf_city_zh` | `这张图片展示了一座现代化城市的美丽景象，画面中显眼的是一座高耸入云的摩天大楼，其外立面在夕阳余晖的映照下显得格外醒目。周围环绕着多栋风格各异的高楼大厦，这些大楼的窗户透出点点灯光，显示出城市的繁华。` |

Baseline command pattern:

```bash
python3 -O examples/offline_inference/text_to_image/text_to_image.py \
  --model /root/autodl-tmp/models/MammothModa2-Preview \
  --stage-configs-path vllm_omni/deploy/mammoth_moda2.yaml \
  --prompt "${PROMPT}" \
  --height 1024 \
  --width 1024 \
  --seed 142 \
  --extra-body '{"text_guidance_scale": 4.0, "cfg_range": [0.0, 1.0], "num_inference_steps": 50}' \
  --output "${OUT_DIR}/${SLUG}_baseline.png"
```

TeaCache command pattern with the final default MammothModa2 coefficients:

```bash
python3 -O examples/offline_inference/text_to_image/text_to_image.py \
  --model /root/autodl-tmp/models/MammothModa2-Preview \
  --stage-configs-path vllm_omni/deploy/mammoth_moda2.yaml \
  --prompt "${PROMPT}" \
  --height 1024 \
  --width 1024 \
  --seed 142 \
  --extra-body '{"text_guidance_scale": 4.0, "cfg_range": [0.0, 1.0], "num_inference_steps": 50}' \
  --cache-backend tea_cache \
  --cache-config '{"rel_l1_thresh":0.1}' \
  --output "${OUT_DIR}/${SLUG}_teacache_default_coeff_t010.png"
```

Coefficient fitting procedure:

```text
1. Run MammothModa2 with TeaCache hook installed but forced full-compute.
2. For every prompt, CFG branch, and adjacent denoising transition, record:
   - input_rel_l1: relative L1 distance between consecutive modulated inputs
   - output_rel_l1: relative L1 distance between consecutive transformer outputs
3. Fit output_rel_l1 = poly(input_rel_l1) with numpy.polyfit(..., deg=4).
```

Fitted MammothModa2 coefficients:

```text
[-972.2069838924724, 493.79503932912513, -70.34862994412688, 6.022465398842022, -0.05835138201804357]
```

Fitting data:

```text
12 train prompts x 2 CFG branches x 49 transitions = 1176 valid points
input_rel_l1 range: 0.02502 - 0.30273
output_rel_l1 range: 0.03687 - 1.08594
degree-4 ordinary least-squares fit: MAE=0.04807, RMSE=0.07688, R^2=0.7972
```

The candidate selection also evaluated an earlier 4-prompt fit and a conservative weighted fit. The 12-prompt ordinary least-squares fit was selected because it had the best train RMSE/R^2 among the non-conservative candidates, matched the 4-prompt fit's TeaCache skip decisions on the 12 calibration prompts at `rel_l1_thresh=0.1`, and matched the 4-prompt fit's quality/speed on the 4 holdout prompts while being fitted from broader prompt coverage.

GPU memory was sampled once per second with:

```bash
nvidia-smi --query-gpu=memory.used --format=csv,noheader,nounits
```

Image similarity was computed between baseline and TeaCache outputs for the same prompt/seed using:

```text
MAE, MSE, RMSE, PSNR, and grayscale SSIM
```

**vLLM Version:** `0.25.0`

**vLLM-Omni Commit:** `8fc3644c99a76358014cf9630fc995e39dcb802b`

### E2E Performance

All final e2e output images were saved successfully at `1024x1024`.

| Prompt | Baseline total | TeaCache total | Total speedup | Baseline DiT stage | TeaCache DiT stage | DiT speedup |
|---|---:|---:|---:|---:|---:|---:|
| `nyc_motorcycle` | `22.9886s` | `18.0485s` | `1.274x` | `22614.98ms` | `17670.17ms` | `1.280x` |
| `cat_laptop` | `23.0409s` | `18.0769s` | `1.275x` | `22618.23ms` | `17703.88ms` | `1.278x` |
| `futuristic_city` | `22.9806s` | `17.4326s` | `1.318x` | `22609.84ms` | `17080.91ms` | `1.324x` |
| `hf_city_zh` | `23.0236s` | `17.2654s` | `1.334x` | `22634.13ms` | `16903.78ms` | `1.339x` |
| **Average** | `23.0084s` | `17.7059s` | `1.300x` | `22619.29ms` | `17339.69ms` | `1.305x` |

### GPU Memory

Peak GPU memory was measured from `nvidia-smi` during each full process run, including model loading and generation.

| Prompt | Baseline peak | TeaCache peak | Delta |
|---|---:|---:|---:|
| `nyc_motorcycle` | `60567 MiB` | `60567 MiB` | `0 MiB` |
| `cat_laptop` | `60567 MiB` | `60567 MiB` | `0 MiB` |
| `futuristic_city` | `60567 MiB` | `60567 MiB` | `0 MiB` |
| `hf_city_zh` | `60567 MiB` | `60567 MiB` | `0 MiB` |
| **Average** | `60567 MiB` | `60567 MiB` | `0 MiB` |

TeaCache did not increase peak GPU memory in this final e2e set. Since the peak includes model loading and runtime allocation behavior, this should be interpreted as "no observed peak-memory regression" rather than a guaranteed memory reduction.

### Output Similarity vs Baseline

Metrics compare the TeaCache output against the no-cache baseline output for the same prompt and seed.

| Prompt | PSNR | SSIM gray | MAE | RMSE |
|---|---:|---:|---:|---:|
| `nyc_motorcycle` | `32.58` | `0.9789` | `2.24` | `5.99` |
| `cat_laptop` | `27.89` | `0.9591` | `5.76` | `10.28` |
| `futuristic_city` | `25.10` | `0.9203` | `5.41` | `14.17` |
| `hf_city_zh` | `22.82` | `0.8148` | `8.50` | `18.42` |
| **Average** | `27.10` | `0.9183` | `5.48` | `12.22` |

Visual inspection confirmed that the fitted MammothModa2 coefficients remove the large composition drift observed with the old Lumina-bootstrap coefficients. The final outputs are not pixel-identical to baseline, but the major subject/layout structure is preserved across the 4-prompt sample.

For PR presentation, the final visual comparison should use the two-column contact sheet:

contact_sheet_default_coeff_threshold_sweep

It shows only `No TeaCache` vs `TeaCache, rel_l1_thresh=0.1`.

### Prompts, Images, and Data Artifacts

Main PR visual artifacts:

The final 4-prompt e2e set used these prompts and generated these images:

| Slug | Prompt |
|---|---|
| `nyc_motorcycle` | `A stylish woman riding a motorcycle through rainy New York City at night, cinematic movie poster style, neon reflections, sharp focus.` |
| `cat_laptop` | `A fluffy orange cat sitting on a laptop keyboard in a cozy home office, warm desk lamp, shallow depth of field, detailed fur.` |
| `futuristic_city` | `A futuristic coastal city at sunset with glass skyscrapers, flying transit vehicles, sailboats on calm water, detailed architecture, soft pink sky.` |
| `hf_city_zh` | `这张图片展示了一座现代化城市的美丽景象，画面中显眼的是一座高耸入云的摩天大楼，其外立面在夕阳余晖的映照下显得格外醒目。周围环绕着多栋风格各异的高楼大厦，这些大楼的窗户透出点点灯光，显示出城市的繁华。` |

### Coefficient Fitting Impact

The old MammothModa2 coefficients were bootstrapped from a Lumina-style reference. With `rel_l1_thresh=0.1`, they accelerated generation but caused large visual drift:

| Coefficients | Avg total speedup | Avg DiT speedup | Avg PSNR | Avg SSIM gray | Avg MAE | Visual result |
|---|---:|---:|---:|---:|---:|---|
| Old Lumina-bootstrap | `1.250x` | `1.255x` | `11.80` | `0.5668` | `47.68` | Large composition drift; `hf_city_zh` split artifact |
| New Mammoth-fitted | `1.300x` | `1.305x` | `27.10` | `0.9183` | `5.48` | No observed large composition drift in the 4-prompt sample |

I also verified that passing only `--cache-config '{"rel_l1_thresh":0.1}'` after the source coefficient update produces the same image as explicitly passing the selected 12-prompt fitted coefficients on `glass_teapot` (`PSNR=inf`, `SSIM=1.0000`, `MAE=0.0000`). This confirms the new default coefficient path is exercised.

Holdout candidate comparison at `rel_l1_thresh=0.1`:

| Candidate | Avg total speedup | Avg DiT speedup | Avg PSNR | Avg SSIM gray | Avg MAE | Notes |
|---|---:|---:|---:|---:|---:|---|
| 4-prompt fitted | `1.309x` | `1.316x` | `29.29` | `0.8711` | `5.20` | Earlier candidate |
| 12-prompt ordinary least squares | `1.310x` | `1.317x` | `29.29` | `0.8711` | `5.20` | Selected; broader calibration set |
| 12-prompt conservative weighted | `1.196x` | `1.200x` | `31.67` | `0.9352` | `3.14` | Better quality, lower speed; not selected as default |

I also reran the final source-default coefficients across the same 4-prompt regression set with `rel_l1_thresh=0.10`, `0.15`, and `0.20`.

| Threshold | Avg total speedup | Avg DiT speedup | Avg PSNR | Avg SSIM gray | Avg MAE | Avg peak memory delta | Visual result |
|---:|---:|---:|---:|---:|---:|---:|---|
| `0.10` | `1.300x` | `1.305x` | `27.10` | `0.9183` | `5.48` | `0.0 MiB` | Best quality/speed tradeoff in this sweep |
| `0.15` | `1.563x` | `1.581x` | `22.45` | `0.8628` | `9.34` | `0.0 MiB` | Faster, but more visible detail/layout drift |
| `0.20` | `1.768x` | `1.793x` | `21.01` | `0.8126` | `12.04` | `-867.5 MiB` | Fastest, but `hf_city_zh` shows substantial layout drift |

Based on this check, `rel_l1_thresh=0.1` is the selected PR value: it keeps the useful `~1.30x` average total speedup while preserving noticeably better output similarity than the more aggressive thresholds.

Per-prompt threshold sweep data:

| Prompt | rel_l1_thresh | Total speedup | DiT speedup | Peak memory delta | PSNR | SSIM gray | MAE |
|---|---:|---:|---:|---:|---:|---:|---:|
| `nyc_motorcycle` | `0.10` | `1.274x` | `1.280x` | `0 MiB` | `32.58` | `0.9789` | `2.24` |
| `nyc_motorcycle` | `0.15` | `1.520x` | `1.537x` | `0 MiB` | `20.18` | `0.8657` | `10.92` |
| `nyc_motorcycle` | `0.20` | `1.729x` | `1.753x` | `-3470 MiB` | `19.64` | `0.8500` | `11.96` |
| `cat_laptop` | `0.10` | `1.275x` | `1.278x` | `0 MiB` | `27.89` | `0.9591` | `5.76` |
| `cat_laptop` | `0.15` | `1.542x` | `1.558x` | `0 MiB` | `26.08` | `0.9420` | `6.98` |
| `cat_laptop` | `0.20` | `1.783x` | `1.805x` | `0 MiB` | `24.78` | `0.9237` | `8.33` |
| `futuristic_city` | `0.10` | `1.318x` | `1.324x` | `0 MiB` | `25.10` | `0.9203` | `5.41` |
| `futuristic_city` | `0.15` | `1.584x` | `1.603x` | `0 MiB` | `22.06` | `0.8722` | `8.92` |
| `futuristic_city` | `0.20` | `1.747x` | `1.775x` | `0 MiB` | `21.40` | `0.8345` | `10.64` |
| `hf_city_zh` | `0.10` | `1.334x` | `1.339x` | `0 MiB` | `22.82` | `0.8148` | `8.50` |
| `hf_city_zh` | `0.15` | `1.608x` | `1.627x` | `0 MiB` | `21.47` | `0.7714` | `10.52` |
| `hf_city_zh` | `0.20` | `1.812x` | `1.839x` | `0 MiB` | `18.22` | `0.6424` | `17.22` |

### Regression Tests

The cache-related regression subset completed under the same local `python3 -O` workaround used for e2e:

```text
python3 -O -m pytest -q \
  tests/diffusion/cache/test_cache_backends.py \
  tests/diffusion/cache/test_teacache_extractors.py

48 passed, 15 warnings in 36.69s
```

Syntax/format checks passed:

```text
python3 -m py_compile \
  vllm_omni/diffusion/cache/teacache/config.py \
  vllm_omni/diffusion/cache/teacache/backend.py \
  vllm_omni/diffusion/cache/teacache/hook.py
```

**BEFORE SUBMITTING:** read [CONTRIBUTING.md](https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md) and run the [precheck-pr skill](https://github.com/vllm-project/vllm-omni/blob/main/.claude/skills/precheck-pr/SKILL.md) with the code agent for a self-check against project conventions.
(anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

File: vllm_omni/model_executor/models/mammoth_moda2/mammoth_moda2.py [hunk 1]
```diff
@@ -738,9 +743,43 @@ def __init__(self, *, vllm_config: VllmConfig, prefix: str = ""):
         else:
             raise ValueError(f"Unsupported model_stage: {self.model_stage}")
 
+        self._dit_cache_backend = None
+        self._maybe_enable_dit_cache_backend()
+
         # Expose intermediate tensor factory for PP if provided by the submodule.
         self.make_empty_intermediate_tensors = getattr(self.model, "make_empty_intermediate_tensors", lambda: None)
 
+    def _maybe_enable_dit_cache_backend(self) -> None:
```
Comment (RuixiangMa) [inline vllm_omni/model_executor/models/mammoth_moda2/mammoth_moda2.py:942]: _maybe_enable_dit_cache_backend re-parses cache_config if it's a string, but _normalize_cache_config already normalizes to a dict upstream.

File: vllm_omni/engine/async_omni_engine.py [hunk 2]
```diff
@@ -1136,6 +1136,45 @@ def _apply_strategy_lb_policy(self, derived: str | None, kwargs: dict[str, Any])
             )
             self._omni_lb_policy = str(derived)
 
+    @classmethod
+    def _inject_diffusion_cache_config(cls, stage_configs: list[Any], kwargs: dict[str, Any]) -> None:
```
Comment (RuixiangMa) [inline vllm_omni/engine/async_omni_engine.py]: enable_cache_dit_summary injection is dead code for MammothModa2 — Silently filtered out by filter_dataclass_kwargs(OmniEngineArgs, ...) since the field isn't on OmniEngineArgs. Not a bug, but misleading

Comment (nagisa-kunhah) [reply vllm_omni/engine/async_omni_engine.py]: Thanks, I remove it now.

File: vllm_omni/model_executor/models/mammoth_moda2/mammoth_moda2.py (hunk 1 above)
Comment (nagisa-kunhah) [reply vllm_omni/model_executor/models/mammoth_moda2/mammoth_moda2.py:942]: Got it. I remove the redundant code now.

File: vllm_omni/diffusion/cache/teacache/extractors.py [hunk 3]
```diff
@@ -819,6 +820,109 @@ def postprocess(h):
...
+def extract_mammoth_moda2_context(
+    module: nn.Module,
+    hidden_states: torch.Tensor,
+    timestep: torch.Tensor,
+    text_hidden_states: torch.Tensor,
+    freqs_cis: torch.Tensor,
+    text_attention_mask: torch.Tensor,
+    ref_image_hidden_states: list[list[torch.Tensor]] | None = None,
+    return_dict: bool = False,
+    teacache_branch: str | None = None,
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/cache/teacache/extractors.py:834]: The positive DiT call always forwards `ar_image_hidden_states` and `ar_image_attention_mask`, but this extractor accepts neither (or `**kwargs`). The TeaCache hook forwards them here and fails before the first denoise step with `TypeError`. Please handle the two AR-image arguments.

File: vllm_omni/diffusion/cache/teacache/extractors.py [hunk 4]
```diff
@@ -819,6 +820,109 @@ def postprocess(h):
...
+
+    batch_size, height, width = module._validate_inputs(
+        hidden_states,
+        text_hidden_states,
+        text_attention_mask,
+        ref_image_hidden_states,
+        return_dict,
+    )
+
+    (
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/cache/teacache/extractors.py:848]: `_prepare_embeddings()` returns 11 items, including `text_attention_mask`, but this tuple unpacks 10. Once the kwargs issue is fixed, this path raises `ValueError: too many values to unpack`. Please include the returned attention mask here.

File: vllm_omni/model_executor/models/mammoth_moda2/mammoth_moda2.py [hunk 5]
```diff
@@ -928,9 +932,37 @@ def __init__(self, *, vllm_config: VllmConfig, prefix: str = ""):
...
+        if self.model_stage != "dit" or self.dit is None:
+            return
+
+        model_config = self.vllm_config.model_config
+        cache_backend_name = getattr(model_config, "cache_backend", None)
+        if cache_backend_name in (None, "", "none"):
+            return
+
+        cache_config = getattr(model_config, "cache_config", None)
```
Comment (hsliuustc0106) [inline vllm_omni/model_executor/models/mammoth_moda2/mammoth_moda2.py]: `cache_config` can still be the raw CLI JSON string after stage runtime overrides. `get_cache_backend()` only converts dicts, then the Mammoth TeaCache enabler dereferences `.rel_l1_thresh`; `--cache-config '{"rel_l1_thresh":0.1}'` raises `AttributeError`. Please normalize it here or at the stage-override boundary.

File: vllm_omni/model_executor/models/mammoth_moda2/mammoth_moda2.py [hunk 6]
```diff
@@ -928,9 +932,37 @@ def __init__(self, *, vllm_config: VllmConfig, prefix: str = ""):
...
+        if cache_config is None:
+            cache_config = {}
+
+        cache_backend = get_cache_backend(cache_backend_name, cache_config)
```
Comment (hsliuustc0106) [inline vllm_omni/model_executor/models/mammoth_moda2/mammoth_moda2.py:975]: This also enables `cache_dit` for Mammoth. CacheDiT falls back to `pipeline.transformer`, while Mammoth exposes only `gen_transformer`, so model initialization crashes. Please whitelist TeaCache here or reject unsupported backends with a clear error.

File: tests/model_executor/models/mammoth_moda2/test_mammoth_moda2_teacache.py [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+        self.timesteps = torch.arange(num_inference_steps, 0, -1, device=device, dtype=torch.float32)
+
+    def step(self, model_pred, t, latents, return_dict=False):  # noqa: ARG002
+        return (latents,)
+
+
+class _FakeTransformer(nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.config = SimpleNamespace(in_channels=4)
```
Comment (hsliuustc0106) [inline tests/model_executor/models/mammoth_moda2/test_mammoth_moda2_teacache.py:28]: These tests never reach their cache assertions: the pipeline reads `self.gen_transformer.time_caption_embed` before the denoise loop, but this fake does not provide it. Add `time_caption_embed` (with `image_embedder=None`) so `_run_pipeline()` exercises the intended branches.

File: vllm_omni/diffusion/cache/teacache/extractors.py (hunk 3 above)
Comment (nagisa-kunhah) [reply vllm_omni/diffusion/cache/teacache/extractors.py:834]: Thanks for your reminder! This may have been caused by rebasing the main code, but I have fixed it now.

File: vllm_omni/diffusion/cache/teacache/extractors.py (hunk 4 above)
Comment (nagisa-kunhah) [reply vllm_omni/diffusion/cache/teacache/extractors.py:848]: Same as above.

File: vllm_omni/model_executor/models/mammoth_moda2/mammoth_moda2.py (hunk 6 above)
Comment (nagisa-kunhah) [reply vllm_omni/model_executor/models/mammoth_moda2/mammoth_moda2.py:975]: Thanks! I added the whitelist here. When we get to `cache_dit`, we can just add it to the same whitelist.

File: tests/model_executor/models/mammoth_moda2/test_mammoth_moda2_teacache.py (hunk 7 above)
Comment (nagisa-kunhah) [reply tests/model_executor/models/mammoth_moda2/test_mammoth_moda2_teacache.py:28]: Thanks! In commit 54981f8, I updated _FakeTransformer to provide time_caption_embed.image_embedder = None, so the pipeline reaches the denoising loop. I also added tests covering the positive and negative TeaCache branches.

File: vllm_omni/model_executor/models/mammoth_moda2/mammoth_moda2.py (hunk 5 above)
Comment (nagisa-kunhah) [reply vllm_omni/model_executor/models/mammoth_moda2/mammoth_moda2.py]: Thanks! Fixed by parsing the stage-level cache_config JSON string into a dictionary before backend initialization.

**Comment by nagisa-kunhah** [comment]
Hi @hsliuustc0106, could you please take another look at this PR? Thanks!

**Comment by Levius-Fubuki** [comment]
Hi @nagisa-kunhah — #7086 moves the MammothModa2 DiT stage from the legacy `LLM_GENERATION` path to the shared `DIFFUSION` runtime.

Because of that ownership change, could you please rebase/refactor #5357 on top of #7086 and integrate TeaCache through the shared diffusion runtime cache hooks? The MammothModa2-specific CFG branch handling, fitted coefficients, and quality/performance evidence can remain valuable, but the cache configuration and enable/refresh lifecycle should no longer be owned by `MammothModa2ForConditionalGeneration` or reintroduced through the legacy generation-stage path.

#7086 intentionally keeps TeaCache out of scope so the two PRs remain separable.

**Comment by nagisa-kunhah** [comment]
> Hi @nagisa-kunhah — #7086 moves the MammothModa2 DiT stage from the legacy `LLM_GENERATION` path to the shared `DIFFUSION` runtime.
> 
> Because of that ownership change, could you please rebase/refactor #5357 on top of #7086 and integrate TeaCache through the shared diffusion runtime cache hooks? The MammothModa2-specific CFG branch handling, fitted coefficients, and quality/performance evidence can remain valuable, but the cache configuration and enable/refresh lifecycle should no longer be owned by `MammothModa2ForConditionalGeneration` or reintroduced through the legacy generation-stage path.
> 
> #7086 intentionally keeps TeaCache out of scope so the two PRs remain separable.

@Levius-Fubuki Sure, Can I rebase on top of pr #7134?

**Comment by Levius-Fubuki** [comment]
Yes, please rebase/stack #5357 on top of #7134. The current head is `17c9bee50f442adbf8a167aa3db9f968fc463df6`.

Please keep #5357 targeting `main` and treat #7134 as its prerequisite. After #7134 is merged, #5357 can be rebased onto the updated `main`.

For the integration boundary, `cache_backend` and `cache_config` should flow through `OmniDiffusionConfig`. The shared `DiffusionModelRunner` should remain responsible for calling the cache backend's `enable()` and per-request `refresh()` methods. MammothModa2-specific changes can stay limited to the TeaCache enabler, CFG positive/negative branch handling, fitted coefficients, and validation evidence.

Please remove the legacy cache plumbing and lifecycle handling from `MammothModa2ForConditionalGeneration`. Note that #7134 currently supports request mode only; step mode and cache-plus-step execution remain out of scope.

#7134 is still Draft pending A800 validation, so I will notify you if that validation requires an interface adjustment.

**Comment by Levius-Fubuki** [comment]
Follow-up on the interface-adjustment question: the A800 validation for #7134 is complete, and it did not require any change to the integration boundary described above.

Please continue to stack #5357 on the shared diffusion runtime: `cache_backend` and `cache_config` should flow through `OmniDiffusionConfig`, while `DiffusionModelRunner` owns cache `enable()` and per-request `refresh()`. MammothModa2-specific code can remain limited to the TeaCache enabler, CFG branch handling, fitted coefficients, and its own quality/performance evidence.

One important branch update: #7134 was rebased and cleaned, so its old head `17c9bee5` was replaced by `68dfe9c1`. If #5357 was already based on the old head, please rebase it onto the refreshed #7134 branch. The current boundary is still request mode only; step execution and cache-plus-step execution remain out of scope.

**Comment by nagisa-kunhah** [comment]
> Follow-up on the interface-adjustment question: the A800 validation for #7134 is complete, and it did not require any change to the integration boundary described above.
> 
> Please continue to stack #5357 on the shared diffusion runtime: `cache_backend` and `cache_config` should flow through `OmniDiffusionConfig`, while `DiffusionModelRunner` owns cache `enable()` and per-request `refresh()`. MammothModa2-specific code can remain limited to the TeaCache enabler, CFG branch handling, fitted coefficients, and its own quality/performance evidence.
> 
> One important branch update: #7134 was rebased and cleaned, so its old head `17c9bee5` was replaced by `68dfe9c1`. If #5357 was already based on the old head, please rebase it onto the refreshed #7134 branch. The current boundary is still request mode only; step execution and cache-plus-step execution remain out of scope.

ok, I will work on this tonight.

**Comment by shihongzhi** [comment]
- This PR is missing the two follow-up fixes #7134 landed after the Sept-8 snapshot it was built from: 9915e54 (VAE output postprocessing; touches pipeline_mammothmoda2_dit.py, diffusion/registry.py) and e9fa8b7 (preserve Mammoth request sampling; touches pipeline_mammothmoda2_dit.py, stage_input_processors/mammoth_moda2.py, deploy/mammoth_moda2.yaml, plus ~97 lines of tests). The sampling one is not cosmetic for this PR: request sampling params (num_inference_steps etc.) are exactly what feeds TeaCache refresh (_refresh_cache_for_requests) and the skip-decision math, so coefficients fitted and benchmarks measured on the pre-fix code may not hold once rebased. Head-to-head, the two core mammoth files differ by only +3/−33 lines vs #7134 (the +3 being the teacache_branch kwargs), so the actual conflict surface is small.
- The recipe's benchmark evidence is currently unreproducible. recipes/MammothModa2/MammothModa2.md:184-185 cites baseline caed3061 and candidate 19de562a, but neither SHA resolves in the repo (orphaned by force-push), and those numbers were measured on the pre-sampling-fix code anyway.
GitHub also reports the PR as unmergeable (conflicts) against current main.

Request: please rebase onto #7134 @ e9fa8b73b and drop the four duplicated migration commits, so this PR contains only the TeaCache changes (it would shrink from 30 files to roughly the 20 TeaCache-specific ones). After the rebase, the speedup claim and the fitted coefficients should be re-measured on the new stack before we can review the quality budget.

**Comment by nagisa-kunhah** [comment]
@shihongzhi Thanks for the reminder! I re-ran this PR, and here are the results:

| Source and setting | Git commit | Cache |
| --- | --- | --- |
| main baseline | `6daf5b30f07cc0b3db87766d22fdf433f8a1a7c9` | disabled |
| current PR | `552005592ab76e4f11b9f2972fe22192dcd38d27` | disabled |
| current PR | `552005592ab76e4f11b9f2972fe22192dcd38d27` | TeaCache, `rel_l1_thresh=0.075` |

The fetched `origin/main` is an ancestor of the branch; the branch contains 16
additional commits. Main and both current-PR settings used BF16, eager mode, a single RTX PRO 6000
Blackwell GPU, CUDNN_ATTN, 1024x1024, 50 inference steps, guidance 4.0, CFG
range `[0, 1]`, and local model revision
`ef5a5e41dbf0de1ef6275586b7580f0d4248b4c6`. Both tested Git trees were clean.
The original branch checkout has the unrelated untracked calibration document.

| Source and setting | E2E mean (s) | E2E stdev (s) | AR mean (s) | DiT mean (s) | DiT stdev (s) | Peak memory (MiB) |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| main baseline | 67.995 | 0.306 | 58.987 | 8.976 | 0.042 | 10,450 |
| current PR, cache disabled | 67.997 | 0.127 | 59.005 | 8.960 | 0.029 | 10,450 |
| current PR, TeaCache 0.075 | 66.156 | 0.138 | 58.511 | 7.620 | 0.107 | 10,618 |

Relative to current main, the branch with TeaCache reduced end-to-end time by
2.705% (1.028x) and DiT-stage time by 15.110% (1.178x). It increased recorded
peak memory by 168 MiB (1.61%). The cache-disabled branch differs from main by
-0.003% E2E and +0.176% DiT, which is within observed measurement variation;
there is no measurable cache-disabled regression in this run.

The two repeats for each setting were pixel-identical. The current-PR cache-disabled images
were also pixel-identical to main. Compared with main, the six TeaCache images
(three distinct images, each repeated) had mean SSIM 0.971458, median SSIM
0.974924, worst SSIM 0.960387, mean LPIPS-SqueezeNet 0.015136, median LPIPS
0.014663, worst LPIPS 0.018148, mean PSNR 34.198 dB, and mean pixel MAE
0.007427.

This pr is ready for review. Feel free to comment here!
